Added support for multiple teleporters per id. Added onEnter and onExit code hooks to the teleports

BAXENdev opened the pull request:
This is still being tested, but I am creating this now so you can look at it today. I won't be back on to finish this until tomorrow (I work graveyard shift).

Multiple Teleporter Support:
* Removed checks for only 2 teleporters.
* Exit teleporter is chosen randomly `exit = random (teleporters - enterTeleporter)`
* Updated deleteAnomalies function

OnEnter and OnExit Code Hooks:
* The teleport module's `Arguments` moved to `Attributes` and `onEnterCode` and `onExitCode` attributes added
* Added `_onEnterCode` and `_onExitCode` params to `createTeleport` function. Module support also added.
* `onEnterCode` and `onExitCode` support added to activateTeleport.

diwako commented on addons/main/functions/anomalies/fn_activateTeleport.sqf:
```diff
@@ -77,7 +76,18 @@ _proxy = _exit getVariable QGVAR(sound);
             if (_doTeleport) then {
                 // [_obj, [((_exitPos select 0) + (random 4) - 2), ((_exitPos select 1) + (random 4) - 2), (_exitPos select 2) ]] remoteExec ["setPos", _obj];
+                
+                try {
+                    [_trg, _exit, _obj] call _onEnterCode;
+                } catch {
+                    hintC ("OnEnter code failed to execute. Affected anomaly at " + str (_pos));
+                };
                 _obj setPos [((_exitPos select 0) + (random 4) - 2), ((_exitPos select 1) + (random 4) - 2), (_exitPos select 2) ];
+                try {
+                    [_trg, _exit, _obj] call _onExitCode;
+                } catch {
+                    hintC ("OnExit code failed to execute. Affected anomaly at " + str (_pos));
+                };
```
Not really a fan of try catch in sqf. Other languages are fine, as they catch actually exceptions, sqf does not.

Here instead of the enter and exit codes why not raise local CBA events?

BAXENdev replied:
I decided that since everything is running on the server, that the code hooks should be server side only too. If you would prefer the code be client side, then I would use CBA events. But in the current implementation, using events is just adding an extra step to calling the code. 

Also, I do not know if raising an event on a local machine is synchronous or not, so I do not know if CBA events would ruin the order of execution. OnEnter is ran before you teleport ("when you enter the teleporter"), you are teleported, and then onExit is ran after you teleport ("when you exit the teleporter").

Try Catch is there to guarantee that bad code does not cause the entire function stop if a player passes bad code. I am not worried about why the code failed to execute; that is on the player to figure out.

That said, if you are fine with it erring out and requiring players to fix it, that is entirely valid to and I will remove the try-catch. But the last bit of code (lines 99-103, timing out the cooldown) will not run in case of an error with onEnter/Exit, which I think causes the teleporter to completely break.

Let me know what you would like or make any changes you prefer.

diwako replied:
The code here is already running on the sever, raising a local event will only raise on the server itself. A client will not receive it at all. In this case a client really does not need the event, as the whole teleport handling is done on the server. It is up to the mission maker to run code on the client if they want to, we just provide the API that something has occurred.

Mission makers can add the event listener to the server and implement their own handling. Error handling is done by CBA then, the teleport code should still continue to run while the mission maker or server receives the script error that happened in the executed custom code. Same mentality as CBA/ACE, if bad custom code is used, let the code creator fix it, it is a Sisyphean endeavor trying to prevent or correcting user errors.

It is way less handling on our end to do then, certainly no need to compile the code from a string or add a try and catch in case bad code has been given.

Plus easier to document on my end in the wiki as well.

diwako commented on addons/main/functions/anomalies/fn_activateTeleport.sqf:
```diff
@@ -30,11 +30,8 @@ if (count _teleporters < 2) exitWith {
-        _exit = _x;
-    };
-} forEach _teleporters;
+_otherTeleporters = _teleporters - [_trg];
+_exit = selectRandom _otherTeleporters;
```
```suggestion
_exit = selectRandom (_teleporters - [_trg]);
```

Why, no need to assign a one time use variable if the content of the variable is short enough.

diwako commented on addons/main/cfgVehicles.hpp:
```diff
@@ -106,12 +114,40 @@ class cfgVehicles {
-                typeName = "NUMBER";
-                defaultValue = 0;
+                tooltip = "$STR_anomaly_teleport_id_desc";
+                property = "anomaly_teleport_id";
```
```suggestion
                property = QGVAR(teleport_id);
```

Tho why need this attribute in the first place?

BAXENdev replied:
This is used when you select multiple objects and edit their attributes all at once. it is used detect that it is the same attribute. I do not use CBA typically, so I had not thought to use QGVAR, but it does make sense to.

diwako commented on addons/main/cfgVehicles.hpp:
```diff
@@ -106,12 +114,40 @@ class cfgVehicles {
+            class onEnterCode {
+                control = "EditCodeMulti5";
+                displayName = "On Enter Code";
+                property = "anomaly_teleport_enterCode";
```
```suggestion
                property = QGVAR(teleport_enterCode);
```

diwako commented on addons/main/cfgVehicles.hpp:
```diff
@@ -106,12 +114,40 @@ class cfgVehicles {
+            class onExitCode {
+                control = "EditCodeMulti5";
+                displayName = "On Exit Code";
+                property = "anomaly_teleport_exitCode";
```
```suggestion
                property = QGVAR(teleport_exitCode);
```

diwako commented:
Just a FYI why the build failed. There is an error in the stringtable.xml. Nothing on your end, it was part of the polish translations. It is fixed on the master branch already, you just need to merge or rebase the changes.

BAXENdev commented:
Actually, I do have a `<br/>` in the stringtable for the attribute descriptions that I forgot to mention. That isnt breaking it the xml, is it?

diwako commented:
> Actually, I do have a `<br/>` tag in the string table for the attribute descriptions that I forgot to mention. That isnt breaking the xml, is it?
> 
> I am the doing the next set of revisions now.

Always escape < and > in stringtable texts. There are a bunch of websites which you can just convert the text to the escaped format.

BAXENdev commented:
Current Status:
* Enter/Exit code moved to teleportOnEnter and teleportOnExit event handlers
* Updated teleporter zen menu minimum ID value to 0 instead of 1
* Module `Arguments` class converted to `Attributes` class
* Updated activate teleport to remove any objNull when it finds a objNull in the teleport list
* Updated function documentation and string table
* Added >2 teleporter support. Random teleporter selected when teleporters >2

Testing is still in progress. Teleport behavior is working.
The updated behavior of attempting to find a new teleporter when a null teleporter is found is still being tested.

diwako commented on addons/main/stringtable.xml:
```diff
@@ -102,12 +102,29 @@
-            <English>Teleporters need the same ID to work together. It is only allowed to have the same ID twice!</English>
+            <Original>Teleporters need the same ID to work together. Several teleporters can have the same ID, but the exit teleporter is randomly selected.</Original>
+            <English>Teleporters need the same ID to work together. Several teleporters can have the same ID, but the exit teleporter is randomly selected.</English>
             <German>Teleporter Anomalien brauchen die selbe ID so dass diese zusammen arbeiten können. Man kann nicht die selbe ID mehr als zwei mal nutzen!</German>
```
```suggestion
            <German>Teleporter Anomalien brauchen die selbe ID so dass diese zusammen arbeiten können. Man kann die selbe ID mehr als zwei mal nutzen!</German>
```
Updated german translation

diwako commented on addons/main/stringtable.xml:
```diff
@@ -102,12 +102,29 @@
+            <Original>Teleporters need the same ID to work together. Several teleporters can have the same ID, but the exit teleporter is randomly selected.</Original>
+            <English>Teleporters need the same ID to work together. Several teleporters can have the same ID, but the exit teleporter is randomly selected.</English>
             <German>Teleporter Anomalien brauchen die selbe ID so dass diese zusammen arbeiten können. Man kann nicht die selbe ID mehr als zwei mal nutzen!</German>
             <Russian>Телепортам требуются одинаковые ID для синхронизации. Можно использовать один и тот же ID только в 2-х аномалиях!</Russian>
```
these are machine translated tho

```suggestion
            <Russian>Телепорты должны иметь одинаковый ID, чтобы работать вместе. Несколько телепортов могут иметь одинаковый ID, но выходной телепорт выбирается случайным образом.</Russian>
```

diwako commented on addons/main/stringtable.xml:
```diff
@@ -102,12 +102,29 @@
+            <English>Teleporters need the same ID to work together. Several teleporters can have the same ID, but the exit teleporter is randomly selected.</English>
             <German>Teleporter Anomalien brauchen die selbe ID so dass diese zusammen arbeiten können. Man kann nicht die selbe ID mehr als zwei mal nutzen!</German>
             <Russian>Телепортам требуются одинаковые ID для синхронизации. Можно использовать один и тот же ID только в 2-х аномалиях!</Russian>
             <Polish>Teleportery potrzebują tego samego ID, by działać. Tylko dwie anomalie mogą mieć to samo ID!</Polish>
```
```suggestion
            <Polish>Teleporty muszą mieć ten sam identyfikator, aby mogły ze sobą współpracować. Kilka teleporterów może mieć ten sam identyfikator, ale teleporter wyjściowy jest wybierany losowo.</Polish>
```

diwako commented on addons/main/functions/anomalies/fn_createTeleport.sqf:
```diff
@@ -8,19 +8,37 @@
+    _onEnterCode = _pos getVariable [QGVAR(onEnterCode), ""];
+    _onExitCode = _pos getVariable [QGVAR(onExitCode), ""];
+
+    try {
+        _onEnterCode = compile _onEnterCode;
+    } catch {
+        _onEnterCode = {};
+        hintC ("On Enter Code failed to compile. Affected anomaly at " + str (getPosASL _pos));
+    };
+    try {
+        _onExitCode = compile _onExitCode;
+    } catch {
+        _onExitCode = {};
+        hintC ("On Exit Code failed to compile. Affected anomaly at " + str (getPosASL _pos));
+    };
```
With the cba events, i doubt these are really needed anymore. A mission maker can easily just add an eventlistener at this point.

BAXENdev replied:
Okay. I will revert it back.
As a personal preference, I like having module specific code. I like that there is a entry point in a module in 3den.

BAXENdev commented:
Current Status:
* Added >2 teleporter support. Random teleporter selected when teleporters >2
* Enter/Exit code completely removed
* Updated teleporter zen menu minimum ID value to 0 instead of 1
* Updated activate teleport to remove any objNull when it finds a objNull in the teleport list
* Updated function documentation and string table (please double check that teleport ID description is correct)
* All anomaly modules (besides Comet and PsyDischarge) have their variable name set to the created trigger upon creation

Tested in multiplayer with 2 local clients (1 client as server, and 1 as client) in 3den multiplayer session.

BAXENdev commented on addons/main/functions/anomalies/fn_createBurner.sqf:
```diff
@@ -18,12 +18,17 @@ params[["_pos",[0,0,0]]];
     _pos set [2,0];
 };
 private _trg = createTrigger ["EmptyDetector", _pos];
+if !(_varName isEqualTo "") then { missionNamespace setVariable [_varName, _trg, true]; };
```
I did not know that existed

BAXENdev commented on addons/main/functions/anomalies/fn_createBurner.sqf:
```diff
@@ -18,12 +18,17 @@ params[["_pos",[0,0,0]]];
+if !(_pos isEqualType []) then {
+    _varName = vehicleVarName _pos;
+    _pos = [_pos] call FUNC(getLocationFromModule);
+};
```
I am going that condition change for you. That requires some consideration for when the function is called by the 3den module vs a regular call. I give modules their own init functions, and I am not quite sure how the parameters are fine here (i am too tired to figure this out now).

diwako replied:
TBH not much need. When the function is called by a module, the param are `[moduleObj, unitArray, acitvateBool]`.

Wrapping things into module functions and calling those functions in the module itself is cleaner, yes, but as long a the modules are this simple no need imo.

unless you want to go really by the book on this one.

diwako commented on addons/main/functions/anomalies/fn_activateTeleport.sqf:
```diff
@@ -29,15 +29,18 @@ if (count _teleporters < 2) exitWith {
 
-if (isNull _exit) exitWith {
-    hintC ("It was not possible to find an exit for teleport anomaly at " + str(getPos _trg) + " with id " + str(_id) + "!")
+if (isNull _exit) then {
```
careful, selectRandom can return a nil value. In that case it will error out.
